src/backend/cuda/copy.cpp

```diff
@@ -22,87 +22,74 @@ using common::is_complex;
-        ptr = out.get();
+void copyData(T *data, const Array<T> &src) {
+    if (src.elements() > 0) {
+        Array<T> lin = src.isReady() && src.isLinear() ? src : copyArray(src);
+        // out is now guaranteed linear
+        auto stream = cuda::getActiveStream();
+        CUDA_CHECK(cudaMemcpyAsync(data, lin.get(), lin.elements() * sizeof(T),
+                                   cudaMemcpyDeviceToHost, stream));
+        CUDA_CHECK(cudaStreamSynchronize(stream));
     }
-
-    auto stream = cuda::getActiveStream();
-    CUDA_CHECK(cudaMemcpyAsync(dst, ptr, src.elements() * sizeof(T),
-                               cudaMemcpyDeviceToHost, stream));
-    CUDA_CHECK(cudaStreamSynchronize(stream));
 }
 
 template<typename T>
 Array<T> copyArray(const Array<T> &src) {
     Array<T> out = createEmptyArray<T>(src.dims());
-    if (src.elements() == 0) { return out; }
-
-    if (src.isLinear()) {
-        CUDA_CHECK(
-            cudaMemcpyAsync(out.get(), src.get(), src.elements() * sizeof(T),
-                            cudaMemcpyDeviceToDevice, cuda::getActiveStream()));
-    } else {
+    if (src.isReady()) {
         kernel::memcopy<T>(out, src, src.ndims());
+    } else {
+        Param<T> info(out.get(), src.dims().dims, src.strides().dims);
+        evalNodes(info, src.getNode().get());
     }
     return out;
 }
 
 template<typename T>
-void multiply_inplace(Array<T> &in, double val) {
-    kernel::copy<T, T>(in, in, in.ndims(), scalar<T>(0), val);
+void multiply_inplace(Array<T> &src, double norm) {
+    kernel::copy<T, T>(src, src, src.ndims(), scalar<T>(0), norm);
 }
 
 template<typename inType, typename outType>
 struct copyWrapper {
-    void operator()(Array<outType> &out, Array<inType> const &in) {
-        kernel::copy<inType, outType>(out, in, in.ndims(), scalar<outType>(0),
-                                      1);
+    void operator()(Array<outType> &dst, Array<inType> const &src) {
+        kernel::copy<inType, outType>(dst, src, src.ndims(), scalar<outType>(0),
+                                      1.0);
     }
 };
 
 template<typename T>
 struct copyWrapper<T, T> {
-    void operator()(Array<T> &out, Array<T> const &in) {
-        if (out.isLinear() && in.isLinear() &&
-            out.elements() == in.elements()) {
-            CUDA_CHECK(cudaMemcpyAsync(
-                out.get(), in.get(), in.elements() * sizeof(T),
-                cudaMemcpyDeviceToDevice, cuda::getActiveStream()));
+    void operator()(Array<T> &dst, Array<T> const &src) {
+        if (dst.isLinear() && src.isLinear() &&
+            dst.elements() == src.elements()) {
+            if (src.isReady()) {
+                CUDA_CHECK(cudaMemcpyAsync(
+                    dst.get(), src.get(), src.elements() * sizeof(T),
+                    cudaMemcpyDeviceToDevice, cuda::getActiveStream()));
+            } else {
+                Param<T> info(dst.get(), src.dims().dims, dst.strides().dims);
+                evalNodes(info, src.getNode().get());
```

Comment:
```suggestion
                evalNodes(dst, src.getNode().get());
```
Array<T> objects implicity convert to Param<T>

Reply:
src.dims() can be different from dst.dims(), so I have to build my own version of Param<T>.

Example: test_assign_cuda.exe